File: go/pkg/memory/compact.go
package memory

import (
	"context"
	"fmt"
	"time"

	"github.com/haivivi/giztoy/go/pkg/recall"
)

// Compact checks all compactable buckets and compacts any that exceed
// the configured [CompressPolicy] thresholds. Compaction cascades:
// compacting the 1h bucket may push segments into 1d, which may then
// also need compaction, and so on up to 1y → lt.
//
// Compact is safe to call frequently — it is a no-op for buckets that
// are within their thresholds. Returns nil if no compaction was needed.
//
// If no compressor is configured, Compact returns nil (no-op).
func (m *Memory) Compact(ctx context.Context) error {
	if m.compressor == nil {
		return nil
	}

	for _, bucket := range recall.CompactableBuckets {
		if err := m.CompactBucket(ctx, bucket); err != nil {
			return fmt.Errorf("memory: compact bucket %s: %w", bucket, err)
		}
	}
	return nil
}

// CompactBucket compacts a single bucket if it exceeds the configured
// thresholds. It reads the oldest segments that push the bucket over
// the threshold, feeds their summaries to [Compressor.CompactSegments],
// stores the result in the appropriate target bucket (determined by time
// span), and deletes the compacted source segments.
//
// The target bucket is chosen by [recall.BucketForSpan]: the smallest
// bucket whose duration covers the time span of the compacted segments.
// This means segments may skip levels (e.g., 1h → 1m) if the source
// segments span a long time range.
func (m *Memory) CompactBucket(ctx context.Context, bucket recall.Bucket) error {
	if m.compressor == nil {
		return nil
	}

	count, chars, err := m.index.BucketStats(ctx, bucket)
	if err != nil {
		return err
	}

	if !m.policy.shouldCompress(chars, count) {
		return nil
	}

	// Read all segments in this bucket (oldest first).
	segments, err := m.index.BucketSegments(ctx, bucket)
	if err != nil {
		return err
	}
	if len(segments) == 0 {
		return nil
	}

	// Determine how many segments to compact: take the oldest ones
	// that bring the bucket back under the threshold.
	// We compact at least half the segments to avoid thrashing.
	compactCount := len(segments) / 2
	if compactCount < 1 {
		compactCount = 1
	}
	// Ensure we leave the bucket under threshold after compaction.
	// If count is over threshold, compact enough to get under.
	if m.policy.MaxMessages > 0 && count > m.policy.MaxMessages {
		need := count - m.policy.MaxMessages + 1
		if need > compactCount {
			compactCount = need
		}
	}
	if compactCount > len(segments) {
		compactCount = len(segments)
	}

	toCompact := segments[:compactCount]

	// Collect summaries for the compressor.
	summaries := make([]string, len(toCompact))
	for i, seg := range toCompact {
		summaries[i] = seg.Summary
	}

	// Run the compactor.
	result, err := m.compressor.CompactSegments(ctx, summaries)
	if err != nil {
		return fmt.Errorf("compact segments: %w", err)
	}

	// Determine target bucket from time span.
	firstTS := toCompact[0].Timestamp
	lastTS := toCompact[len(toCompact)-1].Timestamp
	span := time.Duration(lastTS-firstTS) * time.Nanosecond
	targetBucket := recall.BucketForSpan(span)

	// Ensure the target bucket is coarser than the source.
	// If BucketForSpan returns the same or finer bucket, move to the
	// next coarser one. This prevents segments from staying in the
	// same bucket forever.
	targetBucket = ensureCoarser(bucket, targetBucket)

	// Store the new compacted segments.
	for _, seg := range result.Segments {
		ts := nowNano()
		newSeg := recall.Segment{
			ID:        fmt.Sprintf("%s-%d", m.id, ts),
			Summary:   seg.Summary,
			Keywords:  seg.Keywords,
			Labels:    seg.Labels,
			Timestamp: lastTS, // use the latest timestamp from source segments
			Bucket:    targetBucket,
		}
		if err := m.index.StoreSegment(ctx, newSeg); err != nil {
			return fmt.Errorf("store compacted segment: %w", err)
		}
	}

	// Delete the source segments.
	for _, seg := range toCompact {
		if err := m.index.DeleteSegment(ctx, seg.ID); err != nil {
			return fmt.Errorf("delete source segment %s: %w", seg.ID, err)
		}
	}

	return nil
}

// ensureCoarser ensures the target bucket is strictly coarser than the
// source bucket. If target is the same as or finer than source, it
// returns the next coarser bucket.
func ensureCoarser(source, target recall.Bucket) recall.Bucket {
	buckets := recall.AllBuckets
	sourceIdx := -1
	targetIdx := -1
	for i, b := range buckets {
		if b == source {
			sourceIdx = i
		}
		if b == target {
			targetIdx = i
		}
	}

	if targetIdx <= sourceIdx {
		// Target is same or finer — move to next coarser.
		next := sourceIdx + 1
		if next >= len(buckets) {
			return recall.BucketLT
		}
		return buckets[next]
	}
	return target
}

File: rust/memory/src/compact.rs
use std::time::Duration;

use giztoy_recall::{
    Segment, all_buckets, bucket_for_span, bucket_lt, compactable_buckets, Bucket,
};

use crate::error::MemoryError;
use crate::memory::Memory;
use crate::types::now_nano;

impl Memory {
    /// Check all compactable buckets and compact any exceeding thresholds.
    /// Cascades: compacting 1h may push into 1d, which may also need compaction.
    /// No-op if no compressor is configured.
    pub async fn compact(&self) -> Result<(), MemoryError> {
        if !self.has_compressor() {
            return Ok(());
        }

        for bucket in compactable_buckets() {
            self.compact_bucket(&bucket).await?;
        }
        Ok(())
    }

    /// Compact a single bucket if it exceeds thresholds.
    async fn compact_bucket(&self, bucket: &Bucket) -> Result<(), MemoryError> {
        let compressor = match self.compressor_ref() {
            Some(c) => c,
            None => return Ok(()),
        };

        let (count, chars) = self.index().bucket_stats(bucket)?;
        if !self.policy().should_compress(chars, count) {
            return Ok(());
        }

        let segments = self.index().bucket_segments(bucket)?;
        if segments.is_empty() {
            return Ok(());
        }

        // Compact at least half to avoid thrashing.
        let mut compact_count = segments.len() / 2;
        if compact_count < 1 {
            compact_count = 1;
        }
        if self.policy().max_messages > 0 && count > self.policy().max_messages {
            let need = count - self.policy().max_messages + 1;
            if need > compact_count {
                compact_count = need;
            }
        }
        if compact_count > segments.len() {
            compact_count = segments.len();
        }

        let to_compact = &segments[..compact_count];

        let summaries: Vec<String> = to_compact.iter().map(|s| s.summary.clone()).collect();
        let result = compressor.compact_segments(&summaries).await?;

        let first_ts = to_compact[0].timestamp;
        let last_ts = to_compact[to_compact.len() - 1].timestamp;
        let span = Duration::from_nanos((last_ts - first_ts) as u64);
        let target_bucket = ensure_coarser(bucket, &bucket_for_span(span));

        let mut next_ts = last_ts;
        for (idx, seg_input) in result.segments.into_iter().enumerate() {
            // Segment storage key is keyed by bucket+timestamp.
            // Ensure each compacted segment has a distinct timestamp to avoid
            // key collisions when a compressor returns multiple segments.
            let seg_ts = if idx == 0 {
                last_ts
            } else {
                next_ts
                    .checked_add(1)
                    .unwrap_or_else(now_nano)
            };
            next_ts = seg_ts;

            let id_ts = now_nano();
            let new_seg = Segment {
                id: format!("{}-{}", self.id(), id_ts),
                summary: seg_input.summary,
                keywords: seg_input.keywords,
                labels: seg_input.labels,
                timestamp: seg_ts,
                bucket: target_bucket.as_str().to_string(),
            };
            self.index().store_segment(&new_seg).await?;
        }

        for seg in to_compact {
            self.index().delete_segment(&seg.id)?;
        }

        Ok(())
    }
}

/// Ensure target bucket is strictly coarser than source.
fn ensure_coarser(source: &Bucket, target: &Bucket) -> Bucket {
    let buckets = all_buckets();
    let source_idx = buckets.iter().position(|b| b == source);
    let target_idx = buckets.iter().position(|b| b == target);

    match (source_idx, target_idx) {
        (Some(si), Some(ti)) if ti > si => target.clone(),
        (Some(si), _) => {
            let next = si + 1;
            if next >= buckets.len() {
                bucket_lt()
            } else {
                buckets[next].clone()
            }
        }
        _ => target.clone(),
    }
}
